### ISSUE TITLE
Navigation to source link is broken

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#72638]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Navigation-to-source-link-is-broken/10621259)._

---
[severity:It's more difficult to complete my work] [regression] [worked-in:17.8 (not sure)]
Steps to reproduce:

1. Create a new .NET 8 console app project.
1. Navigate to Console.WriteLine

Expected result: Navigates to sources from source link.
Actual result: Navigates to metadata (only definitions).

Output from the “Navigate to external sources” output window:

> Navigating to symbol 'System.Console.WriteLine(string?)' from 'System.Console'.
Symbol found in assembly path 'C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\8.0.3\ref\net8.0\System.Console.dll'
Symbol found in assembly path 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.3\System.Console.dll'
Cannot find or open the PDB file.
Cannot find or open the PDB file.
Could not find PDB on disk, or embedded, or on a symbol server.

My settings:

Text Editor - C# - Advanced:
![grafik.png](https://aka.ms/dc/image?name=Ba8a2193217f8427a921692ee11515d79638465992260337138_20240321-072705-grafik.png&tid=a8a2193217f8427a921692ee11515d79638465992260337138)

Debugging - General:
![grafik.png](https://aka.ms/dc/image?name=Ba66e4cce6b224126a836348ae171389d638465992823230981_20240321-072803-grafik.png&tid=a66e4cce6b224126a836348ae171389d638465992823230981)

Debugging - Symbols:
![grafik.png](https://aka.ms/dc/image?name=Ba66554f5e7ae4d0b9cb1b39239a47a48638465993135686697_20240321-072834-grafik.png&tid=a66554f5e7ae4d0b9cb1b39239a47a48638465993135686697)

---
#### Feedback Bot on 3/21/2024, 01:30 AM: 

(private comment, text removed)

---
### Original Solutions
(no solutions)

**Comment by dibarbet** [comment]
(Triaging) I'm not able to reproduce myself with similar settings on 17.9.4. 
```
Navigating to symbol 'System.Console.WriteLine(string?)' from 'System.Console'.
Symbol found in assembly path 'C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\8.0.3\ref\net8.0\System.Console.dll'
Symbol found in assembly path 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.3\System.Console.dll'
Found PDB on symbol server.
'/_/src/libraries/System.Console/src/System/Console.cs' found via SourceLink.
```

Does it work if you try it a few times?  IIRC there was an issue if the PDB download took too long we would not be able to navigate there the first time.

**Comment by manuel-pallier** [comment]
@dibarbet No, it doesn't work if I try it multiple times, not even with a wait time of a minute. I'm on a fast and stable internet connection and didn't have problems with SourceLink before (some time ago, when either 17.8 or 17.7 was still the current version).

I've now updated to VS 17.9.5 and the problem remains. Although that was expected from the minimal change log.

Your log says "Found PDB on symbol server.". Do I need to enable a symbol server in the VS options? Because I don't have one enabled (see my screenshots in the MS dev community post - for some reason they aren't showing here on GitHub).
If symbol servers need to be enabled: Is this a new requirement? Because SourceLink worked for me before and I didn't change those settings.

**Comment by dibarbet** [comment]
No, the settings in your screenshots seem to match mine at least.  Moving over to area owner to take a look.  Also tagging @davidwengier in case he knows of additional logs we can collect.

**Comment by davidwengier** [comment]
The message "Cannot find or open the PDB file" is not in Roslyn, but comes from the debugger library we use to download PDBs etc. so I'm guessing its definitely symbol server related. Maybe firewall or some other network layer issues?

@chuckries is there a way to get more info from the debugger here about what might be going on?

**Comment by chuckries** [comment]
The log in the output window is the extent of what we have, I would expect it to have better information, as a failure log would not typically look like this. It looks to me like it's possible we are finding the PDB but then failing to open it. 

@manuel-pallier I wonder if it's possible that you have a cached PDB that is corrupt somehow. You could try clearing all instances of System.Console.dll out of your symbol cache (which should default to %TMP%\SymbolCache based on your screenshot. Then try again.

**Comment by manuel-pallier** [comment]
Deleting %TMP%\SymbolCache (the whole folder) seems to have fixed the problem, thanks!
I'm wondering how it got corrupted. Especially since none of the assemblies which I've tried worked before, so it was not just a single currupted file. And I'm sorry but I didn't create a backup before deleting it, so no way to analyze it now 😢 

Maybe it would be a good idea to show this tip to users in case of PDB read errors?
Or maybe even delete the cache automatically?


Btw, I noticed something only partly related to this issue:
When the download of the PDB file is too slow (which seems to happen even on 200 MBit/s internet connections 😕) the first navigation is to metadata, but further navigations to SourceLink work. So far so expected, but what I noticed is that the output seems to be inconsistent.

Here it only showed an exception and no indication that I should try it again later (but a later try did work correctly):
```
Navigating to symbol 'System.Net.WebRequest' from 'System.Net.Requests'.
Symbol found in assembly path 'C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\8.0.3\ref\net8.0\System.Net.Requests.dll'
Symbol found in assembly path 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.3\System.Net.Requests.dll'
Es sind mehr Daten verfügbar. (Exception from HRESULT: 0x800700EA)
   at Microsoft.VisualStudio.Debugger.BrokeredServices.SymbolLocator.Com.ISymbolLocator159.LocateAnyFormatPdbWithChecksums(PdbInfo159& pdbFile, Object pMetadataProvider, String additionalSearchPaths, UInt32 symbolLoadingOptions)
   at Microsoft.VisualStudio.Debugger.BrokeredServices.SymbolLocator.SymbolLocatorService.<LocateSymbolFileAsync>d__2.MoveNext()
Could not find PDB on disk, or embedded, or on a symbol server.
```

Here it correctly said "Subsequent requests may succeed":
```
Navigating to symbol 'string' from 'System.Runtime'.
Symbol found in assembly path 'C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\8.0.3\ref\net8.0\System.Runtime.dll'
Symbol found in assembly path 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.3\System.Runtime.dll'
Symbol found in assembly path 'C:\Program Files\dotnet\shared\Microsoft.NETCore.App\8.0.3\System.Private.CoreLib.dll'
Timed out trying to download PDB from symbol server. Subsequent requests may succeed.
Es sind mehr Daten verfügbar. (Exception from HRESULT: 0x800700EA)
   at Microsoft.VisualStudio.Debugger.BrokeredServices.SymbolLocator.Com.ISymbolLocator159.LocateAnyFormatPdbWithChecksums(PdbInfo159& pdbFile, Object pMetadataProvider, String additionalSearchPaths, UInt32 symbolLoadingOptions)
   at Microsoft.VisualStudio.Debugger.BrokeredServices.SymbolLocator.SymbolLocatorService.<LocateSymbolFileAsync>d__2.MoveNext()
```

**Comment by chuckries** [comment]
@manuel-pallier thank you for your response, there is a lot of good info here. 

> I'm wondering how it got corrupted.

I am too. It's very hard to know, the actual download logic in this scenario is very straightforward using windows HTTP APIs that are quite old. This could truly be a case of cosmic rays caused a bit to flip in transit. Who knows...

> Maybe it would be a good idea to show this tip to users in case of PDB read errors?

I noticed while looking into this issue that we only have the one error message for "couldn't find or open", perhaps we should make these two distinct log messages.

> Btw, I noticed something only partly related to this issue:

What you are observing is currently by design; if we cannot obtain the PDB from symbol servers fast enough (some hardcoded timeout), we fall back to metadata view but allow the PDB download to continue in the background. It is not uncommon to miss out on Source Link in your first F12 invocation. @davidwengier perhaps it is time to revisit the timeouts and or overall experience here and allow Source Link to have a better chance on first F12.

> Here it only showed an exception and no indication that I should try it again later (but a later try did work correctly):

The exception here is a bug that was introduced recently (I think new for 17.9) that we discovered while investigating this issue. This exception should no longer occur in the next version of VS.
